### PR TITLE
Misc Fixes for Intel Package

### LINKED ISSUE #2516: [BUG] USER-INTEL AVX512 + atom_modify first + nve group integration = dynamics change
**Summary**

When simulating an electrochemical system with a mobile solvent ("group sol") and fixed electrodes, the following combination shows altered dynamics resulting in incorrect physical response to charge electrodes:
- package intel (compiled with -xCORE-AVX512)
- `atom_modify first sol`
- `fix nve sol nve`

This likely generalizes to any situation with a combination of fixed and moving particles, but I have only tested this particular electrode-solvent system in detail.

**LAMMPS Version and Platform**

Issue detected on LAMMPS stable_29Oct2020 (but replicable on master_30Oct2020), compiled with Intel compilers in g++ 8.3.1 mode on CentOS 8, running in serial mode (replicable with MPI). Issue does not replicate when compiled with gcc, and does not replicate when '-xCORE-AVX2' is used (including being forced by removing '-xHost' from cmake/CMakeLists.txt and cmake/Packages/USER-INTEL files). In relation to this issue, compiler warnings about "unroll 4" are seen when compiling VerletLRTIntel.cpp with -xCORE-AVX512.

**Expected Behavior**

For a system with some particles mobile ("group sol") and other particles frozen (using fix setforce 0 0 0), any combination of the following choices should not alter the behavior of the system:

`atom_modify first sol (vs no atom_modify)`
`fix nve sol nve (vs fix nve all nve)`

This should be the case since `atom_modify` only alters the memory ordering of particle information and should not alter dynamics (besides accumulation of floating point errors), while `fix nve` should not change particle behavior upon integrating the entire system vs integrating only unfrozen particles.

**Actual Behavior**

With package intel, compiled with flag -xCORE-AVX512, combining `atom_modify first sol` with `fix nve sol nve` perturbs system dynamics considerably; it appears to severely reduce the effect of interactions between the "sol" group and the rest of the system.

**Steps to Reproduce**

As a practical example, the attached archive contains a small configuration of ionic liquid between two graphene electrodes (~2000 particles), which has been equilibrated with negative charge on the left and positive charge on the right. All ionic liquid particles and graphene particles interact via `lj/cut/coul/long`, and only ionic liquid particles are allowed to move while graphene particles are frozen with `setforce 0 0 0` and not integrating them in `fix nve`. Image below shows visualization of the system (but with two extra layers of graphene, for production runs -- attached archive has only one layer per electrode for speed).

image

At run start the electrode polarities are instantaneously swapped using `set [group] charge ` so that the left electrode is now positive and the right electrode negative, but left constant afterwards. The response of the ionic liquid is tracked in thermo output using the total dipole moment (sum of q*z across all particles).

The correct behavior is for the dipole to increase to about 600 electron-angstroms within 1 ps (500 steps), and this is seen for all combinations of `atom_modify` and `fix nve` without package intel as well as with package intel compiled with -xCORE-AVX2.

But with package intel compiled with -xCORE-AVX512, when both commands are given, the dipole only increases to 300 electron-angstroms within 1 ps. Longer integration times show that the steady state under package intel -xCORE-AVX512 similarly has a smaller dipole (about 0.5x) than is correct. It is difficult to identify exactly what goes wrong -- `fix shake` shows much larger divergences of the ionic liquid molecules from their equilibrium shapes, and yet the system can evolve indefinitely without exploding.

In the archive, inputA has the combination `atom_modify` and `fix nve` and will show erroneous behavior when run with `-pk intel 0 omp 1` (with or without `mode double`) compiled with -xCORE-AVX512, but correct behavior otherwise. inputB has no `atom_modify`, and inputC has `fix nve all nve` instead of `fix nve sol nve`, and both show correct behavior. All inputs run in seconds to a few minutes in serial on a laptop / single core on cluster. The following modifications do not substantially alter the erroneous behavior:

- `fix shake`
- `pair_coeff 5 5 0 0` (turning off LJ interactions on electrode atoms)
- `kspace_modify compute no` (turning off kspace)
- `pair_style lj/cut/coul/long 6.0` (substantially shifting electrostatics from pair to kspace)
- `fix nvt` instead of `fix nve`
- removing `fix setforce 0 0 0` and letting the partial `fix nve` freeze electrode positions
- increasing number of steps

**Further Information, Files, and Links**

The archive of inputs and datafile is here:
[intel_vec_atommodify_nve.tar.gz](https://github.com/lammps/lammps/files/5679610/intel_vec_atommodify_nve.tar.gz)

### LINKED ISSUE #3481: [Feature Request] Script option for _use_table in INTEL PPPM
**Summary**

Enable scripts to turn on/off table-based gridding of charges in INTEL package's PPPM implementation.

**Detailed Description**

Currently Intel's PPPM uses a pre-computed table to spread charge onto the PPPM grid. This setting is currently controlled by `#define INTEL_P3M_TABLE` in `src/INTEL/intel_preprocessor.h`, with the different code paths switched by `if(_use_table)` statements throughout `pppm_intel.cpp` and other similar files.

I would appreciate the option to turn off this tabulation in a user script command (for example `pk intel 0 omp 1 pppm_table off`); it looks like the alternative code paths are already available (including in my `pppm_electrode_intel.cpp`).

I am happy to make the changes myself. I am opening this issue to ask if there is a good reason for this switch not to be user-available.

**Further Information, Files, and Links**

### LINKED ISSUE #3586: [BUG] _LJ/Cut/Intel Incompatable with Ellipsoids
**Summary**
I am trying to run an Intel accelerated simulation of lammps using a hybrid gayberne lj/coul/cut/long potential. After much troubleshooting, I’ve found that the combination of ellipsoids and the LJ/cut/intel potential eventually leads to a particle (seemingly spontaneously) gaining enormous pair energy or thermal energy. The same code with just an LJ/cut (no intel) pair potential runs fine. 

**LAMMPS Version and Platform**
LAMMPS (22 Dec 2022)
Red Hat Enterprise Linux Server 7.9 (Maipo)
Dual 32core Intel Ice Lake 2.6GHz Node
     and
LAMMPS (29 Oct 2020)
CentOS Linux 7 (Core)

**Expected Behavior**
When using LJ/cut, the system runs to completion with appropriate thermostatting.

**Actual Behavior**
When using LJ/cut/intel, the system will eventually (typically in < 2e8 steps) gain large amounts of pair energy and/or kinetic energy (>1e6 in lj units), causing one of the atoms to eventually get lost and the simulation to fail. I have tested langevin damping parameters of 1 & 1000, and several of random seeds with the lj/cut/intel always failing, and the lj/cut never failing. The erratic behavior occurs much sooner (~100,000 steps) for larger systems (1000 atoms).

**Steps to Reproduce**
I run an "sbatch lammpssubmit2.txt" command. It runs inp.txt which reads the datafile data.txt, and outputs out.txt

**Further Information, Files, and Links**
[data.txt](https://github.com/lammps/lammps/files/10338360/data.txt)
[inp.txt](https://github.com/lammps/lammps/files/10338361/inp.txt)
[lammpssubmit2.txt](https://github.com/lammps/lammps/files/10338362/lammpssubmit2.txt)
[out.txt](https://github.com/lammps/lammps/files/10338473/out.txt)

### LINKED ISSUE #3743: [BUG] invalid memory allocation with INTEL package and more than 1 million of atoms per CPU
**Summary**

I'm running large simulations (>20 millions of atoms) with the INTEL package and I see an allocation error with a **negative number**. I suspect there is some integer overflow in the allocation code. As far as I know, 20M atoms is not over any of the [LAMMPS limits](https://docs.lammps.org/Build_settings.html#size-of-lammps-integer-types-and-size-limits), is it?

`ERROR on proc 17: Failed to allocate -8532973184 bytes for array _list_alloc (src/memory.cpp:66)`

**LAMMPS Version and Platform**

lmp-28Mar2023, Linux x86_64, Compiler: Intel Classic C++ 20.21.9 / Intel(R) C++ g++ 11.2.1 mode

```
export CFLAGS="-Ofast -xCOMMON-AVX512" 
export CXXFLAGS=$CFLAGS
cmake -C ../cmake/presets/intel.cmake ../cmake -DPKG_MEAM=yes -DPKG_MANYBODY=yes -DPKG_MC=yes -DPKG_REPLICA=yes -DPKG_OPENMP=yes -DPKG_INTEL=yes -DBUILD_MPI=yes -DBUILD_OMP=yes -DCMAKE_BUILD_TYPE=Release
```

Then I have to remove the flag `-xHost`, which overrides my  `-xCOMMON-AVX512`. It  appeared for some reason automatically in `CMakeFiles/lammps.dir/flags.make` and `CMakeFiles/lmp.dir/flags.make`.

```$ icpc --version
icpc: remark #10441: The Intel(R) C++ Compiler Classic (ICC) is deprecated and will be removed from product release in the second half of 2023. The Intel(R) oneAPI DPC++/C++ Compiler (ICX) is the recommended compiler moving forward. Please transition to use this compiler. Use '-diag-disable=10441' to disable this message.
icpc (ICC) 2021.9.0 20230302
Copyright (C) 1985-2023 Intel Corporation.  All rights reserved.
```

**Expected Behavior**

I prepared a simple script, which demonstrates the problem:
```
processors * * * grid numa
units            metal
variable         size equal 169
boundary         p p p
lattice          fcc 1.0
region           box block 0 ${size} 0 ${size} 0 ${size}
create_box       1 box
create_atoms     1 box
mass             1 1.0
pair_style       lj/cut 2.5
pair_coeff       * * 1 1
run              1
```
When you run the script with size 169 or less (corresponds to 19.31M atoms), it is fine:

```
$ OMP_NUM_THREADS=1 mpirun -n 20 ~/bin/lmp-28Mar2023-intel-intel-2 -sf intel -in test.in
LAMMPS (28 Mar 2023)
  using 1 OpenMP thread(s) per MPI task
Lattice spacing in x,y,z = 1 1 1
Created orthogonal box = (0 0 0) to (169 169 169)
  2 by 2 by 5 MPI processor grid
  2 by 2 by 5 core grid within node
Created 19307236 atoms
  using lattice units in orthogonal box = (0 0 0) to (169 169 169)
  create_atoms CPU = 0.145 seconds
WARNING: No fixes with time integration, atoms won't move (src/verlet.cpp:60)
----------------------------------------------------------
Using INTEL Package without Coprocessor.
Compiler: Intel Classic C++ 20.21.9 / Intel(R) C++ g++ 11.2.1 mode
SIMD compiler directives: Enabled
Precision: mixed
----------------------------------------------------------
  generated 0 of 0 mixed pair_coeff terms from geometric mixing rule
Neighbor list info ...
  update every 1 steps, delay 10 steps, check yes
  max neighbors/atom: 2000, page size: 100000
  master list distance cutoff = 4.5
  ghost atom cutoff = 4.5
  binsize = 2.25, bins = 76 76 76
  1 neighbor lists, perpetual/occasional/extra = 1 0 0
  (1) pair lj/cut/intel, perpetual
      attributes: half, newton on, intel
      pair build: half/bin/newton/intel
      stencil: half/bin/3d
      bin: intel
Setting up Verlet run ...
  Unit style    : metal
  Current step  : 0
  Time step     : 0.001
Per MPI rank memory allocation (min/avg/max) = 8386 | 8462 | 8511 Mbytes
Step Temp E_pair E_mol TotEng Press 
       0            0 2.5560444e+10            0 2.5560444e+10 3.6872569e+10 
       1            0 2.5560444e+10            0 2.5560444e+10 3.6872569e+10 
Loop time of 2.95457 on 20 procs for 1 steps with 19307236 atoms

Performance: 0.029 ns/day, 820.714 hours/ns, 0.338 timesteps/s
99.7% CPU use with 20 MPI tasks x 1 OpenMP threads

MPI task timing breakdown:
Section |  min time  |  avg time  |  max time  |%varavg| %total
---------------------------------------------------------------
Pair    | 2.8131     | 2.8577     | 2.8996     |   1.4 | 96.72
Neigh   | 0          | 0          | 0          |   0.0 |  0.00
Comm    | 0.030073   | 0.068336   | 0.11565    |   9.6 |  2.31
Output  | 0.0087768  | 0.0099657  | 0.011599   |   0.8 |  0.34
Modify  | 0.0073469  | 0.010156   | 0.012226   |   1.6 |  0.34
Other   |            | 0.008396   |            |       |  0.28

Nlocal:         965362 ave      971074 max      956793 min
Histogram: 8 0 0 0 0 0 0 0 0 12
Nghost:         550896 ave      562999 max      545326 min
Histogram: 4 12 0 0 0 0 0 0 0 4
Neighs:    7.25952e+08 ave 7.31034e+08 max 7.18744e+08 min
Histogram: 6 2 0 0 0 0 0 0 3 9

Total # of neighbors = 1.4519041e+10
Ave neighs/atom = 752
Neighbor list builds = 0
Dangerous builds = 0
Total wall time: 0:00:10
```

**Actual Behavior**

When you run the script with size 170 or more (corresponds to 19.65M atoms), you'll see the allocation error:

```
$ OMP_NUM_THREADS=1 mpirun -n 20 ~/bin/lmp-28Mar2023-intel-intel-2 -sf intel -in test.in 
LAMMPS (28 Mar 2023)
  using 1 OpenMP thread(s) per MPI task
Lattice spacing in x,y,z = 1 1 1
Created orthogonal box = (0 0 0) to (170 170 170)
  2 by 2 by 5 MPI processor grid
  2 by 2 by 5 core grid within node
Created 19652000 atoms
  using lattice units in orthogonal box = (0 0 0) to (170 170 170)
  create_atoms CPU = 0.137 seconds
WARNING: No fixes with time integration, atoms won't move (src/verlet.cpp:60)
----------------------------------------------------------
Using INTEL Package without Coprocessor.
Compiler: Intel Classic C++ 20.21.9 / Intel(R) C++ g++ 11.2.1 mode
SIMD compiler directives: Enabled
Precision: mixed
----------------------------------------------------------
  generated 0 of 0 mixed pair_coeff terms from geometric mixing rule
Neighbor list info ...
  update every 1 steps, delay 10 steps, check yes
  max neighbors/atom: 2000, page size: 100000
  master list distance cutoff = 4.5
  ghost atom cutoff = 4.5
  binsize = 2.25, bins = 76 76 76
  1 neighbor lists, perpetual/occasional/extra = 1 0 0
  (1) pair lj/cut/intel, perpetual
      attributes: half, newton on, intel
      pair build: half/bin/newton/intel
      stencil: half/bin/3d
      bin: intel
Setting up Verlet run ...
  Unit style    : metal
  Current step  : 0
  Time step     : 0.001
ERROR on proc 17: Failed to allocate -8532973184 bytes for array _list_alloc (src/memory.cpp:66)
Last command: run              1
Abort(1) on node 17 (rank 17 in comm 0): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 17
```

**Steps to Reproduce**

When running large sizes, but without `-sf intel`, there is no error. So problem is most probably in the INTEL package.

The binary executable is linked to libtbbmalloc.so.2, if it is important.

**Further Information, Files, and Links**

Interesting is, the error shows if I run it on 1-20 CPUs, but does not show on 21-40 CPUs. I.e. `mpirun -n 20` fails, while `mpirun -n 40` does not fail immediately. It fails later on with:

```
ERROR on proc 6: Non-numeric atom coords - simulation unstable (src/OPENMP/domain_omp.cpp:58)
Last command: minimize ${etol} ${ftol} ${maxiter} ${maxeval}
ERROR on proc 9: Non-numeric atom coords - simulation unstable (src/OPENMP/domain_omp.cpp:58)
Last command: minimize ${etol} ${ftol} ${maxiter} ${maxeval}
ERROR on proc 25: Non-numeric atom coords - simulation unstable (src/OPENMP/domain_omp.cpp:58)
Last command: minimize ${etol} ${ftol} ${maxiter} ${maxeval}
Abort(1) on node 9 (rank 9 in comm 0): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 9
ERROR on proc 2: Non-numeric atom coords - simulation unstable (src/OPENMP/domain_omp.cpp:58)
Last command: minimize ${etol} ${ftol} ${maxiter} ${maxeval}
ERROR on proc 5: Non-numeric atom coords - simulation unstable (src/OPENMP/domain_omp.cpp:58)
Last command: minimize ${etol} ${ftol} ${maxiter} ${maxeval}
ERROR on proc 22: Non-numeric atom coords - simulation unstable (src/OPENMP/domain_omp.cpp:58)
Last command: minimize ${etol} ${ftol} ${maxiter} ${maxeval}
ERROR on proc 23: Non-numeric atom coords - simulation unstable (src/OPENMP/domain_omp.cpp:58)
Last command: minimize ${etol} ${ftol} ${maxiter} ${maxeval}
ERROR on proc 35: Non-numeric atom coords - simulation unstable (src/OPENMP/domain_omp.cpp:58)
Last command: minimize ${etol} ${ftol} ${maxiter} ${maxeval}
Abort(1) on node 2 (rank 2 in comm 0): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 2
```

So I believe the memory access is wrong.

### LINKED ISSUE #3784: [BUG] atom/swap accepts all moves when using the intel package
**Summary**

All atom swaps performed by the fix `atom/swap` are accepted in violation of the metropolis criteria, when using the Intel package.

**LAMMPS Version and Platform**

Version: 23 Jun 2022 - Update 3
OS: Linux
CPU: Intel i5-12500

**Further Information, Files, and Links**

[IntelSwapTest.txt](https://github.com/lammps/lammps/files/11504018/IntelSwapTest.txt)
[Swap_withIntel.log](https://github.com/lammps/lammps/files/11504020/Swap_withIntel.log)
[Swap_withoutIntel.log](https://github.com/lammps/lammps/files/11504021/Swap_withoutIntel.log)

Attached are an input script as well as log files from running the script with the following commands:

With Intel: `OMP_NUM_THREADS=1 mpirun -np 4 lmp -in IntelSwapTest.txt -log Swap_withIntel.log -suffix intel -package intel 0 omp 0`

Without Intel: `OMP_NUM_THREADS=1 mpirun -np 4 lmp -in IntelSwapTest.txt -log Swap_withoutIntel.log`

### PR DESCRIPTION
**Pull request by wmbrownIntel**
**Summary**

Small changes to fix outstanding issues and feature requests for Intel package.

- If pre_reverse() is not called as part of force calculation (e.g. MC), the sync for Intel buffers with native arrays will occur in post_force().
- bigint is now used in checks for neighbor overflow allowing for large simulation sizes per MPI task
- Runtime control of the table option for P3M is now available via a package option
- Fixes issue when atom styles using torque are mixed with intel pair styles not using torque
- Fixes issue where use of intel package could erroneously re-enable atom sort causing issues with "atom_modify first"
- Fixes issue recently introduced into long-range thread mode for P3M.
- std:: namespace used for transcendentals to guarantee overloaded functions available.

**Related Issue(s)**

- Fixes #2516
- Fixes #3586
- Fixes #3743
- Fixes #3784
- Implements and thus closes #3481

**Author(s)**

Mike Brown, Intel

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

**Implementation Notes**

**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

### CODE REVIEW

File: src/MC/fix_widom.cpp [hunk 1]
```diff
@@ -1050,13 +1050,7 @@ double FixWidom::energy_full()
 
   if (force->kspace) force->kspace->compute(eflag,vflag);
 
-  // unlike Verlet, not performing a reverse_comm() or forces here
-  // b/c Widom does not care about forces
-  // don't think it will mess up energy due to any post_force() fixes
-  // but Modify::pre_reverse() is needed for INTEL
-
-  if (modify->n_pre_reverse) modify->pre_reverse(eflag,vflag);
-  if (modify->n_pre_force) modify->pre_force(vflag);
```
Comment (akohlmey) [inline src/MC/fix_widom.cpp:1059]: Why remove pre_force? Only pre_reverse was added for INTEL package compatibility in 2e7e5aeac4b90ad13d034c34771e170863ad532e

Comment (wmbrownIntel) [reply src/MC/fix_widom.cpp:1059]: Why are we calling pre_force twice here (in the original code block)?

Comment (athomps) [reply src/MC/fix_widom.cpp:1059]: I think `fix_widom.cpp` is wrong. `modify->pre_force(vflag);` should be `modify->post_force(vflag);`, that is what we have in `fix_gcmc.cpp`

Comment (akohlmey) [reply src/MC/fix_widom.cpp:1059]: @evoyiatzis Do you have any comments on this?

Comment (akohlmey) [reply src/MC/fix_widom.cpp:1059]: @athomps I am a bit confused by your statement. When I compare the `full_energy()` code following the "clear forces" comment, it is now - that is, after the edits by @wmbrownIntel - *identical* to the same section in fix gcmc.

It seems that we need to update the log files for fix widom in the examples section.

**Review by stanmoore1 — APPROVED** [review]
I approve

**Pull request merged**